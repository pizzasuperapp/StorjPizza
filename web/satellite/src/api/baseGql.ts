// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import gql from 'graphql-tag';

import { apollo } from '@/utils/apollo';
import { GraphQLError } from "graphql";

/**
 * BaseGql is a graphql utility which allows to perform queries and mutations.
 */
export class BaseGql {
    /**
     * performs qraphql query.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async query(query: string, variables: any = null): Promise<any> { // eslint-disable-line @typescript-eslint/no-explicit-any
        const response = await apollo.query(
            {
                query: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    /**
     * performs qraphql mutation.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async mutate(query: string, variables: any = null): Promise<any> { // eslint-disable-line @typescript-eslint/no-explicit-any
        const response = await apollo.mutate(
            {
                mutation: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    private combineErrors(gqlError: readonly GraphQLError[]): string {
        return gqlError.map(err => err).join('\n');
    }
}
