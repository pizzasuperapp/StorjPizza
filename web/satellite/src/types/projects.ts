// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all project-related functionality.
 */
export interface ProjectsApi {
    /**
     * Creates project.
     *
     * @param createProjectFields - contains project information
     * @throws Error
     */
    create(createProjectFields: ProjectFields): Promise<Project>;
    /**
     * Fetch projects.
     *
     * @returns Project[]
     * @throws Error
     */
    get(): Promise<Project[]>;
    /**
     * Update project name and description.
     *
     * @param projectId - project ID
     * @param updateProjectFields - project fields to update
     * @param updateProjectLimits - project limits to update
     * @returns Project[]
     * @throws Error
     */
    update(projectId: string, updateProjectFields: ProjectFields, updateProjectLimits: ProjectLimits): Promise<void>;
    /**
     * Delete project.
     *
     * @param projectId - project ID
     * @throws Error
     */
    delete(projectId: string): Promise<void>;

    /**
     * Get project limits.
     *
     * @param projectId- project ID
     * throws Error
     */
    getLimits(projectId: string): Promise<ProjectLimits>;

    /**
     * Get project limits.
     *
     * throws Error
     */
    getTotalLimits(): Promise<ProjectLimits>;

    /**
     * Get project daily usage by specific date range.
     *
     * throws Error
     */
    getDailyUsage(projectID: string, start: Date, end: Date): Promise<ProjectsStorageBandwidthDaily>;

    /**
     * Fetch owned projects.
     *
     * @returns ProjectsPage
     * @throws Error
     */
    getOwnedProjects(cursor: ProjectsCursor): Promise<ProjectsPage>;
}

/**
 * MAX_NAME_LENGTH defines maximum amount of symbols for project name.
 */
export const MAX_NAME_LENGTH = 20;

/**
 * MAX_DESCRIPTION_LENGTH defines maximum amount of symbols for project description.
 */
export const MAX_DESCRIPTION_LENGTH = 100;

/**
 * Project is a type, used for creating new project in backend.
 */
export class Project {
    public constructor(
        public id: string = '',
        public name: string = '',
        public description: string = '',
        public createdAt: string = '',
        public ownerId: string = '',
        public isSelected: boolean = false,
        public memberCount: number = 0,
    ) {}

    /**
     * Returns created date as a local string.
     */
    public createdDate(): string {
        const createdAt = new Date(this.createdAt);
        return createdAt.toLocaleString('en-US', { year: 'numeric', month: '2-digit', day: 'numeric' });
    }
}

/**
 * ProjectFields is a type, used for creating and updating project.
 */
export class ProjectFields {
    public constructor(
        public name: string = '',
        public description: string = '',
        public ownerId: string = '',
    ) {}

    /**
     * checkName checks if project name is valid.
     */
    public checkName(): void {
        this.nameIsNotEmpty();
        this.nameHasLessThenTwentySymbols();
    }

    /**
     * nameIsNotEmpty checks if project name is not empty.
     */
    private nameIsNotEmpty(): void {
        if (this.name.length === 0) throw new Error('Project name can\'t be empty!');
    }

    /**
     * nameHasLessThenTwentySymbols checks if project name has less then 20 symbols.
     */
    private nameHasLessThenTwentySymbols(): void {
        if (this.name.length > MAX_NAME_LENGTH) throw new Error('Name should be less than 21 character!');
    }
}

/**
 * ProjectLimits is a type, used for describing project limits.
 */
export class ProjectLimits {
    public constructor(
        public bandwidthLimit: number = 0,
        public bandwidthUsed: number = 0,
        public storageLimit: number = 0,
        public storageUsed: number = 0,
        public objectCount: number = 0,
        public segmentCount: number = 0,
    ) {}
}

export class ProjectPage {
    public constructor(
        public projects: Project[] = [],
        public pageCount: number = 0,
        public currentPage: number = 0,
        public totalCount: number = 0,
    ) {}
}

/**
 * ProjectsPage class, used to describe paged projects list.
 */
export class ProjectsPage {
    public constructor(
        public projects: Project[] = [],
        public limit: number = 0,
        public offset: number = 0,
        public pageCount: number = 0,
        public currentPage: number = 0,
        public totalCount: number = 0,
    ) {}
}

/**
 * ProjectsPage class, used to describe paged projects list.
 */
export class ProjectsCursor {
    public constructor(
        public limit: number = 0,
        public page: number = 0,
    ) {}
}

/**
 * DataStamp is storage/bandwidth usage stamp for satellite at some point in time
 */
export class DataStamp {
    public constructor(
        public value = 0,
        public intervalStart = new Date()
    ) {}

    /**
     * Creates new empty instance of stamp with defined date
     * @param date - holds specific date of the date range
     * @returns Stamp - new empty instance of stamp with defined date
     */
    public static emptyWithDate(date: Date): DataStamp {
        return new DataStamp(0, date);
    }
}

/**
 * ProjectsStorageBandwidthDaily is used to describe project's daily storage and bandwidth usage.
 */
export class ProjectsStorageBandwidthDaily {
    public constructor(
        public storage: DataStamp[] = [],
        public allocatedBandwidth: DataStamp[] = [],
        public settledBandwidth: DataStamp[] = [],
    ) {}
}

/**
 * ProjectUsageDateRange is used to describe project's usage by date range.
 */
export interface ProjectUsageDateRange {
    since: Date;
    before: Date;
}
