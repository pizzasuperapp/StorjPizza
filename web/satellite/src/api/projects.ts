// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import {
    DataStamp,
    Project,
    ProjectFields,
    ProjectLimits,
    ProjectsApi,
    ProjectsCursor,
    ProjectsPage,
    ProjectsStorageBandwidthDaily,
} from '@/types/projects';
import { HttpClient } from '@/utils/httpClient';
import { Time } from "@/utils/time";

export class ProjectsApiGql extends BaseGql implements ProjectsApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/projects';

    /**
     * Creates project.
     *
     * @param projectFields - contains project information
     * @throws Error
     */
    public async create(projectFields: ProjectFields): Promise<Project> {
        const query =
            `mutation($name: String!, $description: String!) {
                createProject(
                    input: {
                        name: $name,
                        description: $description,
                    }
                ) {id}
            }`;

        const variables = {
            name: projectFields.name,
            description: projectFields.description,
        };

        const response = await this.mutate(query, variables);

        return new Project(response.data.createProject.id, variables.name, variables.description, '', projectFields.ownerId);
    }

    /**
     * Fetch projects.
     *
     * @returns Project[]
     * @throws Error
     */
    public async get(): Promise<Project[]> {
        const query = `query {
            myProjects{
                name
                id
                description
                createdAt
                ownerId
            }
        }`;

        const response = await this.query(query);

        return response.data.myProjects.map((project: Project) => {
            return new Project(
                project.id,
                project.name,
                project.description,
                project.createdAt,
                project.ownerId,
            );
        });
    }

    /**
     * Update project name and description.
     *
     * @param projectId - project ID
     * @param projectFields - project fields
     * @param projectLimits - project limits
     * @returns Project[]
     * @throws Error
     */
    public async update(projectId: string, projectFields: ProjectFields, projectLimits: ProjectLimits): Promise<void> {
        const query =
            `mutation($projectId: String!, $name: String!, $description: String!, $storageLimit: String!, $bandwidthLimit: String!) {
                updateProject(
                    id: $projectId,
                    projectFields: {
                        name: $name,
                        description: $description,
                    },
                    projectLimits: {
                        storageLimit: $storageLimit,
                        bandwidthLimit: $bandwidthLimit,
                    }
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
            name: projectFields.name,
            description: projectFields.description,
            storageLimit: projectLimits.storageLimit.toString(),
            bandwidthLimit: projectLimits.bandwidthLimit.toString(),
        };

        await this.mutate(query, variables);
    }

    /**
     * Delete project.
     *
     * @param projectId - project ID
     * @throws Error
     */
    public async delete(projectId: string): Promise<void> {
        const query =
            `mutation($projectId: String!) {
                deleteProject(
                    id: $projectId
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
        };

        await this.mutate(query, variables);
    }

    /**
     * Get project limits.
     *
     * @param projectId- project ID
     * throws Error
     */
    public async getLimits(projectId: string): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/${projectId}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
                limits.objectCount,
                limits.segmentCount,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get usage limits');
    }

    /**
     * Get total limits for all the projects that user owns.
     *
     * throws Error
     */
    public async getTotalLimits(): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get total usage limits');
    }

    /**
     * Get project daily usage for specific date range.
     *
     * @param projectId- project ID
     * @param start- since date
     * @param end- before date
     * throws Error
     */
    public async getDailyUsage(projectId: string, start: Date, end: Date): Promise<ProjectsStorageBandwidthDaily> {
        const since = Time.toUnixTimestamp(start).toString();
        const before = Time.toUnixTimestamp(end).toString();
        const path = `${this.ROOT_PATH}/${projectId}/daily-usage?from=${since}&to=${before}`;
        const response = await this.http.get(path);

        if (response.ok) {
            const usage = await response.json();

            return new ProjectsStorageBandwidthDaily(
                usage.storageUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.allocatedBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.settledBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get project daily usage');
    }

    /**
     * Fetch owned projects.
     *
     * @returns ProjectsPage
     * @throws Error
     */
    public async getOwnedProjects(cursor: ProjectsCursor): Promise<ProjectsPage> {
        const query =
            `query($limit: Int!, $page: Int!) {
                ownedProjects( cursor: { limit: $limit, page: $page } ) {
                    projects {
                        id,
                        name,
                        ownerId,
                        description,
                        createdAt,
                        memberCount
                    },
                    limit,
                    offset,
                    pageCount,
                    currentPage,
                    totalCount
                 }
             }`;

        const variables = {
            limit: cursor.limit,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.getProjectsPage(response.data.ownedProjects);
    }

    /**
     * Method for mapping projects page from json to ProjectsPage type.
     *
     * @param page anonymous object from json
     */
    private getProjectsPage(page: any): ProjectsPage { // eslint-disable-line @typescript-eslint/no-explicit-any
        if (!page) {
            return new ProjectsPage();
        }

        const projects: Project[] = page.projects.map(key =>
            new Project(
                key.id,
                key.name,
                key.description,
                key.createdAt,
                key.ownerId,
                false,
                key.memberCount));

        return new ProjectsPage(projects, page.limit, page.offset, page.pageCount, page.currentPage, page.totalCount);
    }

}
