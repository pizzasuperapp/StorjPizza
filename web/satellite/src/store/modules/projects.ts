// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    DataStamp,
    Project,
    ProjectFields,
    ProjectLimits,
    ProjectsApi,
    ProjectsCursor,
    ProjectsPage,
    ProjectsStorageBandwidthDaily,
    ProjectUsageDateRange,
} from '@/types/projects';
import {StoreModule} from "@/types/store";

export const PROJECTS_ACTIONS = {
    FETCH: 'fetchProjects',
    FETCH_OWNED: 'fetchOwnedProjects',
    FETCH_DAILY_DATA: 'fetchDailyData',
    CREATE: 'createProject',
    CREATE_DEFAULT_PROJECT: 'createDefaultProject',
    SELECT: 'selectProject',
    UPDATE_NAME: 'updateProjectName',
    UPDATE_DESCRIPTION: 'updateProjectDescription',
    UPDATE_STORAGE_LIMIT: 'updateProjectStorageLimit',
    UPDATE_BANDWIDTH_LIMIT: 'updateProjectBandwidthLimit',
    DELETE: 'deleteProject',
    CLEAR: 'clearProjects',
    GET_LIMITS: 'getProjectLimits',
    GET_TOTAL_LIMITS: 'getTotalLimits',
};

export const PROJECTS_MUTATIONS = {
    ADD: 'CREATE_PROJECT',
    REMOVE: 'DELETE_PROJECT',
    UPDATE_PROJECT_NAME: 'UPDATE_PROJECT_NAME',
    UPDATE_PROJECT_DESCRIPTION: 'UPDATE_PROJECT_DESCRIPTION',
    UPDATE_PROJECT_STORAGE_LIMIT: 'UPDATE_STORAGE_LIMIT',
    UPDATE_PROJECT_BANDWIDTH_LIMIT: 'UPDATE_BANDWIDTH_LIMIT',
    SET_PROJECTS: 'SET_PROJECTS',
    SELECT_PROJECT: 'SELECT_PROJECT',
    CLEAR_PROJECTS: 'CLEAR_PROJECTS',
    SET_LIMITS: 'SET_PROJECT_LIMITS',
    SET_TOTAL_LIMITS: 'SET_TOTAL_LIMITS',
    SET_PAGE_NUMBER: 'SET_PAGE_NUMBER',
    SET_PAGE: 'SET_PAGE',
    SET_DAILY_DATA: 'SET_DAILY_DATA',
    SET_CHARTS_DATE_RANGE: 'SET_CHARTS_DATE_RANGE',
};

const defaultSelectedProject = new Project('', '', '', '', '', true, 0);

export class ProjectsState {
    public projects: Project[] = [];
    public selectedProject: Project = defaultSelectedProject;
    public currentLimits: ProjectLimits = new ProjectLimits();
    public totalLimits: ProjectLimits = new ProjectLimits();
    public cursor: ProjectsCursor = new ProjectsCursor();
    public page: ProjectsPage = new ProjectsPage();
    public allocatedBandwidthChartData: DataStamp[] = [];
    public settledBandwidthChartData: DataStamp[] = [];
    public storageChartData: DataStamp[] = [];
    public chartDataSince: Date = new Date();
    public chartDataBefore: Date = new Date();
}

interface ProjectsContext {
    state: ProjectsState
    commit: (string, ...unknown) => void
    rootGetters: {
        user: {
            id: string
        }
    }
    dispatch: (string, ...unknown) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
}

const {
    FETCH,
    FETCH_DAILY_DATA,
    CREATE,
    CREATE_DEFAULT_PROJECT,
    SELECT,
    UPDATE_NAME,
    UPDATE_DESCRIPTION,
    UPDATE_STORAGE_LIMIT,
    UPDATE_BANDWIDTH_LIMIT,
    DELETE,
    CLEAR,
    GET_LIMITS,
    GET_TOTAL_LIMITS,
    FETCH_OWNED,
} = PROJECTS_ACTIONS;

const {
    ADD,
    REMOVE,
    UPDATE_PROJECT_NAME,
    UPDATE_PROJECT_DESCRIPTION,
    UPDATE_PROJECT_STORAGE_LIMIT,
    UPDATE_PROJECT_BANDWIDTH_LIMIT,
    SET_PROJECTS,
    SELECT_PROJECT,
    CLEAR_PROJECTS,
    SET_LIMITS,
    SET_TOTAL_LIMITS,
    SET_PAGE_NUMBER,
    SET_PAGE,
    SET_DAILY_DATA,
    SET_CHARTS_DATE_RANGE,
} = PROJECTS_MUTATIONS;
const projectsPageLimit = 7;

export function makeProjectsModule(api: ProjectsApi): StoreModule<ProjectsState, ProjectsContext> {
    return {
        state: new ProjectsState(),
        mutations: {
            [ADD](state: ProjectsState, createdProject: Project): void {
                state.projects.push(createdProject);
            },
            [SET_PROJECTS](state: ProjectsState, projects: Project[]): void {
                state.projects = projects;

                if (!state.selectedProject.id) {
                    return;
                }

                const projectsCount = state.projects.length;

                for (let i = 0; i < projectsCount; i++) {
                    const project = state.projects[i];

                    if (project.id !== state.selectedProject.id) {
                        continue;
                    }

                    state.selectedProject = project;

                    return;
                }

                state.selectedProject = defaultSelectedProject;
            },
            [SELECT_PROJECT](state: ProjectsState, projectID: string): void {
                const selected = state.projects.find((project: Project) => project.id === projectID);

                if (!selected) {
                    return;
                }

                state.selectedProject = selected;
            },
            [UPDATE_PROJECT_NAME](state: ProjectsState, fieldsToUpdate: ProjectFields): void {
                state.selectedProject.name = fieldsToUpdate.name;
            },
            [UPDATE_PROJECT_DESCRIPTION](state: ProjectsState, fieldsToUpdate: ProjectFields): void {
                state.selectedProject.description = fieldsToUpdate.description;
            },
            [UPDATE_PROJECT_STORAGE_LIMIT](state: ProjectsState, limitsToUpdate: ProjectLimits): void {
                state.currentLimits.storageLimit = limitsToUpdate.storageLimit;
            },
            [UPDATE_PROJECT_BANDWIDTH_LIMIT](state: ProjectsState, limitsToUpdate: ProjectLimits): void {
                state.currentLimits.bandwidthLimit = limitsToUpdate.bandwidthLimit;
            },
            [REMOVE](state: ProjectsState, projectID: string): void {
                state.projects = state.projects.filter(project => project.id !== projectID);

                if (state.selectedProject.id === projectID) {
                    state.selectedProject = new Project();
                }
            },
            [SET_LIMITS](state: ProjectsState, limits: ProjectLimits): void {
                state.currentLimits = limits;
            },
            [SET_TOTAL_LIMITS](state: ProjectsState, limits: ProjectLimits): void {
                state.totalLimits = limits;
            },
            [CLEAR_PROJECTS](state: ProjectsState): void {
                state.projects = [];
                state.selectedProject = defaultSelectedProject;
                state.currentLimits = new ProjectLimits();
                state.totalLimits = new ProjectLimits();
                state.storageChartData = [];
                state.allocatedBandwidthChartData = [];
                state.settledBandwidthChartData = [];
                state.chartDataSince = new Date();
                state.chartDataBefore = new Date();
            },
            [SET_PAGE_NUMBER](state: ProjectsState, pageNumber: number) {
                state.cursor.page = pageNumber;
                state.cursor.limit = projectsPageLimit;
            },
            [SET_PAGE](state: ProjectsState, page: ProjectsPage) {
                state.page = page;
            },
            [SET_DAILY_DATA](state: ProjectsState, payload: ProjectsStorageBandwidthDaily) {
                state.allocatedBandwidthChartData = payload.allocatedBandwidth;
                state.settledBandwidthChartData = payload.settledBandwidth;
                state.storageChartData = payload.storage;
            },
            [SET_CHARTS_DATE_RANGE](state: ProjectsState, payload: ProjectUsageDateRange) {
                state.chartDataSince = payload.since;
                state.chartDataBefore = payload.before;
            },
        },
        actions: {
            [FETCH]: async function ({commit}: ProjectsContext): Promise<Project[]> {
                const projects = await api.get();

                commit(SET_PROJECTS, projects);

                return projects;
            },
            [FETCH_OWNED]: async function ({commit, state}: ProjectsContext, pageNumber: number): Promise<ProjectsPage> {
                commit(SET_PAGE_NUMBER, pageNumber);

                const projectsPage: ProjectsPage = await api.getOwnedProjects(state.cursor);
                commit(SET_PAGE, projectsPage);

                return projectsPage;
            },
            [FETCH_DAILY_DATA]: async function ({commit, state}: ProjectsContext, payload: ProjectUsageDateRange): Promise<void> {
                const usage: ProjectsStorageBandwidthDaily = await api.getDailyUsage(state.selectedProject.id, payload.since, payload.before)

                commit(SET_CHARTS_DATE_RANGE, payload)
                commit(SET_DAILY_DATA, usage);
            },
            [CREATE]: async function ({commit}: ProjectsContext, createProjectFields: ProjectFields): Promise<Project> {
                const project = await api.create(createProjectFields);

                commit(ADD, project);

                return project;
            },
            [CREATE_DEFAULT_PROJECT]: async function ({rootGetters, dispatch}: ProjectsContext): Promise<void> {
                const UNTITLED_PROJECT_NAME = 'My First Project';
                const UNTITLED_PROJECT_DESCRIPTION = '___';
                const project = new ProjectFields(
                    UNTITLED_PROJECT_NAME,
                    UNTITLED_PROJECT_DESCRIPTION,
                    rootGetters.user.id,
                );
                const createdProject = await dispatch(CREATE, project);

                await dispatch(SELECT, createdProject.id);
            },
            [SELECT]: function ({commit}: ProjectsContext, projectID: string): void {
                commit(SELECT_PROJECT, projectID);
            },
            [UPDATE_NAME]: async function ({commit, state}: ProjectsContext, fieldsToUpdate: ProjectFields): Promise<void> {
                const project = new ProjectFields(
                    fieldsToUpdate.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_NAME, fieldsToUpdate);
            },
            [UPDATE_DESCRIPTION]: async function ({commit, state}: ProjectsContext, fieldsToUpdate: ProjectFields): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    fieldsToUpdate.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_DESCRIPTION, fieldsToUpdate);
            },
            [UPDATE_STORAGE_LIMIT]: async function ({commit, state}: ProjectsContext, limitsToUpdate: ProjectLimits): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    limitsToUpdate.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_STORAGE_LIMIT, limitsToUpdate);
            },
            [UPDATE_BANDWIDTH_LIMIT]: async function ({commit, state}: ProjectsContext, limitsToUpdate: ProjectLimits): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    limitsToUpdate.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_BANDWIDTH_LIMIT, limitsToUpdate);
            },
            [DELETE]: async function ({commit}: ProjectsContext, projectID: string): Promise<void> {
                await api.delete(projectID);

                commit(REMOVE, projectID);
            },
            [GET_LIMITS]: async function ({commit}: ProjectsContext, projectID: string): Promise<ProjectLimits> {
                const limits = await api.getLimits(projectID);

                commit(SET_LIMITS, limits);

                return limits;
            },
            [GET_TOTAL_LIMITS]: async function ({commit}: ProjectsContext): Promise<ProjectLimits> {
                const limits = await api.getTotalLimits();

                commit(SET_TOTAL_LIMITS, limits);

                return limits;
            },
            [CLEAR]: function({commit}: ProjectsContext): void {
                commit(CLEAR_PROJECTS);
            },
        },
        getters: {
            projects: (state: ProjectsState): Project[] => {
                return state.projects.map((project: Project) => {
                    if (project.id === state.selectedProject.id) {
                        project.isSelected = true;
                    }

                    return project;
                });
            },
            projectsWithoutSelected: (state: ProjectsState): Project[] => {
                return state.projects.filter((project: Project) => {
                    return project.id !== state.selectedProject.id;
                });
            },
            selectedProject: (state: ProjectsState): Project => state.selectedProject,
            projectsCount: (state: ProjectsState, rootGetters: ProjectsContext["rootGetters"]): number => {
                let projectsCount = 0;

                state.projects.forEach((project: Project) => {
                    if (project.ownerId === rootGetters.user.id) {
                        projectsCount++;
                    }
                });

                return projectsCount;
            },
        },
    };
}
