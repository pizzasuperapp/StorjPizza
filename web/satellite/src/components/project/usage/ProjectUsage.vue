// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="project-usage">
        <UsageArea
            class="project-usage__storage-used"
            title="Storage"
            :used="storageUsed"
            :limit="storageLimit"
            :is-data-fetching="isDataFetching"
        />
        <UsageArea
            class="project-usage__bandwidth-used"
            title="Bandwidth"
            :used="bandwidthUsed"
            :limit="bandwidthLimit"
            :is-data-fetching="isDataFetching"
        />
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import UsageArea from '@/components/project/usage/UsageArea.vue';

import { PROJECTS_ACTIONS } from '@/store/modules/projects';

// @vue/component
@Component({
    components: {
        UsageArea,
    },
})
export default class ProjectUsage extends Vue {
    public isDataFetching = true;

    /**
     * Lifecycle hook after initial render.
     * Fetches project limits.
     */
    public async mounted(): Promise<void> {
        if (!this.$store.getters.selectedProject.id) {
            return;
        }

        try {
            await this.$store.dispatch(PROJECTS_ACTIONS.GET_LIMITS, this.$store.getters.selectedProject.id);

            this.isDataFetching = false;
        } catch (error) {
            await this.$notify.error(error.message);
        }
    }

    /**
     * Returns bandwidth used amount from store.
     */
    public get bandwidthUsed(): number {
        return this.$store.state.projectsModule.currentLimits.bandwidthUsed;
    }

    /**
     * Returns bandwidth limit amount from store.
     */
    public get bandwidthLimit(): number {
        return this.$store.state.projectsModule.currentLimits.bandwidthLimit;
    }

    /**
     * Returns storage used amount from store.
     */
    public get storageUsed(): number {
        return this.$store.state.projectsModule.currentLimits.storageUsed;
    }

    /**
     * Returns storage limit amount from store.
     */
    public get storageLimit(): number {
        return this.$store.state.projectsModule.currentLimits.storageLimit;
    }
}
</script>

<style scoped lang="scss">
    .project-usage {
        width: 100%;
        font-family: 'font_regular', sans-serif;
        border-radius: 6px;
        display: flex;
        align-items: center;
        justify-content: space-between;

        &__storage-used,
        &__bandwidth-used {
            width: calc(50% - 20px);
        }

        &__storage-used {
            margin-right: 40px;
        }
    }
</style>
