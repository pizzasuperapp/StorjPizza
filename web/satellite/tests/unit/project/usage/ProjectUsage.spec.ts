// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import ProjectUsage from '@/components/project/usage/ProjectUsage.vue';

import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { Project, ProjectLimits } from '@/types/projects';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { ProjectsApiMock } from '../../mock/api/projects';

const localVue = createLocalVue();
localVue.use(Vuex);

const projectLimits = new ProjectLimits(1000, 100, 1000, 100);
const projectsApi = new ProjectsApiMock();
projectsApi.setMockLimits(projectLimits);
const projectsModule = makeProjectsModule(projectsApi);

const store = new Vuex.Store({ modules: { projectsModule }});
const project = new Project('id', 'test', 'test', 'test', 'ownedId', false);

describe('ProjectUsage.vue', () => {
    it('renders correctly', (): void => {
        store.commit(PROJECTS_MUTATIONS.ADD, project);
        store.commit(PROJECTS_MUTATIONS.SELECT_PROJECT, project.id);
        store.commit(PROJECTS_MUTATIONS.SET_LIMITS, projectLimits);

        const wrapper = shallowMount(ProjectUsage, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });
});
