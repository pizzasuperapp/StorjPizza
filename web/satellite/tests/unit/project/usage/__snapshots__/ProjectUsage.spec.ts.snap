// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectUsage.vue renders correctly 1`] = `
<div class="project-usage">
  <usagearea-stub title="Storage" used="100" limit="1000" isdatafetching="true" class="project-usage__storage-used"></usagearea-stub>
  <usagearea-stub title="Bandwidth" used="100" limit="1000" isdatafetching="true" class="project-usage__bandwidth-used"></usagearea-stub>
</div>
`;
